File: suites/06_gpu_compute/comparisons/ocudu_5g_phy.cpp
#include <iostream>
#include <vector>
#include <cstdlib>
#include <algorithm>

int64_t ldpc_decode_layer(int64_t length, int64_t iterations) {
    const int64_t lcg_m = 1664525;
    const int64_t lcg_c = 1013904223;
    const int64_t mod_byte = 256;
    const int64_t mid_val = 128;

    int64_t syndrome_sum = 0;
    for (int64_t iter = 0; iter < iterations; ++iter) {
        for (int64_t i = 0; i < length - 4; i += 4) {
            int64_t l0 = (((i * lcg_m) + lcg_c) % mod_byte) - mid_val;
            int64_t l1 = ((((i + 1) * lcg_m) + lcg_c) % mod_byte) - mid_val;
            int64_t l2 = ((((i + 2) * lcg_m) + lcg_c) % mod_byte) - mid_val;
            int64_t l3 = ((((i + 3) * lcg_m) + lcg_c) % mod_byte) - mid_val;

            int64_t a0 = std::abs(l0);
            int64_t a1 = std::abs(l1);
            int64_t a2 = std::abs(l2);
            int64_t a3 = std::abs(l3);

            int64_t check_val = std::min({a0, a1, a2, a3});
            int64_t sign = ((l0 ^ l1 ^ l2 ^ l3) < 0) ? -1 : 1;

            syndrome_sum += check_val * sign;
        }
    }
    return syndrome_sum;
}

int main() {
    int64_t pdsch_throughput = ldpc_decode_layer(16384, 8);
    std::cout << pdsch_throughput << std::endl;
    return 0;
}

File: suites/06_gpu_compute/comparisons/ocudu_5g_phy.rs
fn ldpc_decode_layer(length: i64, iterations: i64) -> i64 {
    let lcg_m = 1664525i64;
    let lcg_c = 1013904223i64;
    let mod_byte = 256i64;
    let mid_val = 128i64;

    let mut syndrome_sum = 0i64;
    for _iter in 0..iterations {
        let mut i = 0i64;
        while i < length - 4 {
            let l0 = (((i * lcg_m) + lcg_c) % mod_byte) - mid_val;
            let l1 = ((((i + 1) * lcg_m) + lcg_c) % mod_byte) - mid_val;
            let l2 = ((((i + 2) * lcg_m) + lcg_c) % mod_byte) - mid_val;
            let l3 = ((((i + 3) * lcg_m) + lcg_c) % mod_byte) - mid_val;

            let a0 = l0.abs();
            let a1 = l1.abs();
            let a2 = l2.abs();
            let a3 = l3.abs();

            let check_val = a0.min(a1).min(a2).min(a3);
            let sign = if (l0 ^ l1 ^ l2 ^ l3) < 0 { -1 } else { 1 };

            syndrome_sum += check_val * sign;
            i += 4;
        }
    }
    syndrome_sum
}

fn main() {
    let pdsch_throughput = ldpc_decode_layer(16384, 8);
    println!("{}", pdsch_throughput);
}
